src/Compilers/CSharp/Portable/Binder/InMethodBinder.cs

```diff
@@ -38,13 +38,29 @@ internal sealed class InMethodBinder : LocalScopeBinder
 #endif
 
         public InMethodBinder(MethodSymbol owner, Binder enclosing)
-            : base(enclosing, enclosing.Flags & ~BinderFlags.AllClearedAtExecutableCodeBoundary)
+            : base(enclosing, ConstructFlags(owner, enclosing))
         {
             Debug.Assert(!enclosing.Flags.Includes(BinderFlags.InCatchFilter));
             Debug.Assert((object)owner != null);
             _methodSymbol = owner;
         }
 
+        private static BinderFlags ConstructFlags(MethodSymbol owner, Binder enclosing)
+        {
+            BinderFlags flags = enclosing.Flags & ~BinderFlags.AllClearedAtExecutableCodeBoundary;
+
+            if (owner.IsIterator && enclosing.Compilation.IsFeatureEnabled(MessageID.IDS_FeatureRefUnsafeInIteratorAsync))
+            {
+                // Spec §13.3.1: "An iterator block always defines a safe context,
+                // even when its declaration is nested in an unsafe context."
+                // Note that if the iterator method itself has the unsafe modifier,
+                // its body binder will have the UnsafeRegion flag applied by the BinderFactory.
```

Comment:
>// its body binder will have the UnsafeRegion flag applied by the BinderFactory. [](http://example.com/codeflow?start=16&length=80)

I could be wrong, but it looks to me that this assumption is wrong for accessors.

Also, a binder for a local function is not created by a BinderFactory and it looks like LocalBinderFactory doesn't behave according to this assumption.

Reply:
In general, this approach feel fragile and incomplete to me. For example, we are creating other kinds of binders for a method/property. If we are saying that iterator method/properties are "safe" by default, then why we are adjusting only `InMethodBinder`? What about binders that used to bind signatures, etc.?

Given that, I would like to suggest to revert changes made in this file and instead deal with removal of the flag in the same places in code that add it when `unsafe` modifier is explicitly present on the declaration.